### PR TITLE
Track global_amax for weight FP4 MSE sweep; Refactor to NVFP4StaticQantizer, NVFP4MSECalibrator

### PR DESCRIPTION
**Pull request by realAsma**
## What does this PR do?

**Type of change:** ? 

**Overview:** ?

## Usage

```python
# Add a code snippet demonstrating how to use this
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Release Notes

* **New Features**
  * Added NVFP4StaticQuantizer for improved 4-bit quantization with enhanced precision control
  * Introduced NVFP4MSECalibrator with flexible candidate generation for calibration optimization

* **Improvements**
  * Optimized GPU kernels for Hopper+ graphics cards with better performance
  * Extended Triton support to broader GPU compatibility
  * Enhanced backward compatibility for restoring previously quantized models

* **Tests**
  * Added comprehensive test coverage for new quantizers and calibration methods

### CODE REVIEW

File: modelopt/torch/quantization/calib/mse.py [hunk 1]
```diff
@@ -219,3 +153,32 @@ def compute_amax(self, verbose: bool = False):
...
+        amax: torch.Tensor,  # per_block_amax shape [num_blocks]
+        global_amax: torch.Tensor,  # scalar
+        axis: int | tuple | list | None = None,
+        quant_func: Callable | None = None,
+        error_func: Callable | None = None,
+    ):
+        """Initialize NVFP4 MSE calibrator with per-block and global amax."""
+        super().__init__(amax=amax, axis=axis, quant_func=quant_func, error_func=error_func)
+        self._global_amax = global_amax
+
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/calib/mse.py:188]: fixed

File: modelopt/torch/quantization/conversion.py [hunk 2]
```diff
@@ -135,6 +136,25 @@ def restore_quantizer_state(model: nn.Module, config: QuantizeConfig, metadata:
...
+    # Convert quantizers that were saved as NVFP4StaticQuantizer
+    for name, module in model.named_modules():
+        if isinstance(module, TensorQuantizer) and not isinstance(module, NVFP4StaticQuantizer):
+            unwrapped_name = get_unwrapped_name(name, model)
+            if unwrapped_name in quantizer_state_dict:
+                state = quantizer_state_dict[unwrapped_name]
+                if state.get("_is_nvfp4_static_quantizer"):
+                    NVFP4StaticQuantizer.from_tensor_quantizer(module)
+
+    return restore_quantizer_state(model, config, metadata)
```
Comment (Fridah-nv) [inline modelopt/torch/quantization/conversion.py]: I'm trying to understand the logic here: 
When this function is called, the model should contain quantizers and `restore_quantizer_state` will load correct amax values from the checkpoint. We currently do not dispatch `TensorQuantizer` to `NVFP4StaticQuantizer` when the model is initialized as quantized model, so this `convert to NVFP4StaticQuantizer` logic is needed here?

Comment (realAsma) [reply modelopt/torch/quantization/conversion.py]: > We currently do not dispatch TensorQuantizer to NVFP4StaticQuantizer when the model is initialized as quantized model

Yes.

File: modelopt/torch/quantization/triton/fp4_kernel_hopper.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    tile_reshaped = tl.reshape(tile, (TILE_M, NUM_FP4_BLOCKS, BLOCK_SIZE))
+    x_abs = tl.abs(tile_reshaped)
+
+    block_max = tl.max(x_abs, axis=2, keep_dims=True)
+
+    block_max_scaled = block_max / (6.0 * global_scale_safe)
+    block_max_scaled = tl.minimum(block_max_scaled, 448.0)
+    block_max_quant = block_max_scaled.to(tl.float8e4nv).to(tl.float32) * global_scale
+    block_max_quant = tl.where(block_max_quant >= 1e-5, block_max_quant, 1.0)
```
Comment (realAsma) [inline modelopt/torch/quantization/triton/fp4_kernel_hopper.py:85]: @mxinO I am thinking we should make the scale masking consistent with our CUDA FP4 fake quant kernel  - https://github.com/NVIDIA/Model-Optimizer/blob/452c5a09b03f8db34a39151aae4f7a4b01efd9cd/modelopt/torch/quantization/src/tensor_quant_mx.cu#L161

File: modelopt/torch/quantization/calib/mse.py [hunk 4]
```diff
@@ -219,3 +177,32 @@ def compute_amax(self, verbose: bool = False):
                 )
 
         return self._amax
+
+
+class NVFP4MSECalibrator(MseCalibrator):
```
Comment (sugunav14) [inline modelopt/torch/quantization/calib/mse.py:174]: Do we only search for per block scales? What about the global scale?

**Review by sugunav14 — APPROVED** [review]
I reviewed the calibrator and the static quantizer logic! Just had a general question about scale search.

File: modelopt/torch/quantization/triton/fp4_kernel_hopper.py (hunk 3 above)
Comment (mxinO) [reply modelopt/torch/quantization/triton/fp4_kernel_hopper.py:85]: Sure, we can change the masking logic here.

File: modelopt/torch/quantization/calib/mse.py [hunk 5]
```diff
@@ -219,3 +177,32 @@ def compute_amax(self, verbose: bool = False):
...
+        self,
+        amax: torch.Tensor,  # per_block_amax shape [num_blocks]
+        global_amax: torch.Tensor,  # scalar
+        axis: int | tuple | list | None = None,
+        quant_func: Callable | None = None,
+        error_func: Callable | None = None,
+    ):
+        """Initialize NVFP4 MSE calibrator with per-block and global amax."""
+        super().__init__(amax=amax, axis=axis, quant_func=quant_func, error_func=error_func)
+        self._global_amax = global_amax
```
Comment (mxinO) [inline modelopt/torch/quantization/calib/mse.py:187]: I assume the block amaxs would change after MSE searching, so we should re-calculate the global amax after searching, it should help with with quant error of fp8 scales.

Comment (mxinO) [reply modelopt/torch/quantization/calib/mse.py:187]: Sounds the MSE searching should use dynamic global amax.

Comment (realAsma) [reply modelopt/torch/quantization/calib/mse.py:187]: No — in the current design we keep the global amax frozen and do not recompute it after the MSE search, because the search itself is performed under that fixed global scaling.

The MSE search is selecting candidates from FP8-quantized values produced with this frozen global amax/scale. If we update the global amax afterward, we effectively change the FP8 quantization scale, which means the FP8 values the search evaluated are no longer the ones we’ll use at inference/training time. I need to think through whether/how to make that consistent before switching to a “dynamic global amax” approach.

File: modelopt/torch/quantization/calib/mse.py (hunk 4 above)
Comment (realAsma) [reply modelopt/torch/quantization/calib/mse.py:174]: global scale is frozen

File: modelopt/torch/quantization/triton/fp4_kernel_hopper.py (hunk 3 above)
Comment (realAsma) [reply modelopt/torch/quantization/triton/fp4_kernel_hopper.py:85]: done
